language/stdlib/modules/validator_set.mvir

```diff
@@ -105,7 +105,8 @@ module ValidatorSet {
     // Trigger a reconfiguation the Libra system by:
     // (1) Computing a new validator set and storing it on chain
     // (2) Emitting an event containing new validator set, which will be passed to the executor
-    reconfigure() acquires T {
+    // TODO: make  this private + only callable from block prologue
+    public reconfigure() acquires T {
```

Comment:
We can create a `Reconfigurator` resource and store that in the `Block.T`. By doing that we can make sure it is callable only from block prologue without breaking the verifier.

Reply:
That would work (as would merging the `Block` + `ValidatorSet` modules). I'm not sure what the best approach would be yet.

- Merging modules would make the modules ugly, but would let us store the `Block` resource and `ValidatorSet` resource as separate top-level resources/at separate addresses. There are several advantages to this.
- The `Reconfigurator` approach would let us keep the modules nice, but would (I think) force us to combine the validator set and block resources. This would make updating the time (which we need to do every block) expensive because we would need to write back the entire validator set as well (even if it hasn't changed).

My plan is to punt on this decision until we know more about what the block prologue and validator set change logic will look like. I think the right move will become clearer at that point.

Reply:
Why the second approach will force us to combine validator set and block resources? IMHO the `Reconfigurator` resource should be an empty resource that acts likes a `MintCapability`.

Reply:
Ah, I think I didn't understand the proposal. Were you suggesting something like:

```
module Block {
  resource Reconconfigerator { }
  prologue() {
    ValidatorSet.reconfigure(Reconfigerator {});
  }
```

If so, that wouldn't work because it creates a circular dependency. `ValidatorSet` would need to know about the `Block.ReconfigeratorType`. We could add the `Reconfigerator` type in `ValidatorSet` instead, but then we couldn't ensure that only `Block.prologue` can access it.

Reply:
Ah I was indeed think of your second proposal. I was thinking of doing something similar as `MintCapability` where we have a private function that creates the unique instance of that resource and store it in `Block`

Reply:
Yeah, unfortunately the `MintCapability` trick only works because no one outside of `LibraCoin` needs to see that type.